### PR TITLE
Preliminary LLVM 5 port (not as the default LLVM)

### PR DESCRIPTION
**Pull request by dmk42**
This change ports Chapel to LLVM 5.  Two issues remain to be addressed, as documented in #7246 .
* LLVM 5 requires `CHPL_LLVM_DEVELOPER` to be set.
* Using `--llvm --fast` causes `release/examples/benchmerks/hpcc/fft` to fail with anomalous numeric results, along with some other fft tests.

Submitting this change now allows setting aside the remaining issues until after the Chapel 1.16 release, and also allows others to try LLVM 5 without having to duplicate my work.

I expect the items in #6993 to be addressed after the Chapel 1.16 release.

The primary changes to the LLVM API are:
* LLVM has changed more of its API to use named variadic parameters, which allows a function with a variable number of arguments to handle the exact number of arguments safely without requiring a null-terminated argument list.  Therefore, `NULL` as an argument must now be deleted from several method calls.
* The BBVectorize pass has been fully deleted in favor of SLP vectorization.
* Some methods now take an extra argument to specify an address space into which allocation should occur.
* `SynchronizationScope` has been renamed `SyncScope::ID`.
* Call site attributes are now stored in an `AttributeList` instead of an `AttributeSet`.

Testing (all done with `CHPL_LLVM_DEVELOPER` set due to the issue noted above):
- [x] Passes `release/examples` and `llvm` on a Mac without `--fast`
- [x] Passes `release/examples` and `llvm` on a Mac with `--fast` except for `fft` as noted above.
- [x] Passes the same tests on Ubuntu except for the casting issue noted in #6231
- [x] Full local linux64 testing without `--fast`
- [x] Full local linux64 testing with `--fast` (failing `fft` as noted above)
- [x] LLVM 4 verified to still work after this change (full local linux64 testing with and without `--fast`, and Mac testing on `release/examples` and `llvm`)

### CODE REVIEW

**Comment by dmk42** [comment]
@mppf - Would you mind taking a look at this?

File: compiler/codegen/type.cpp [hunk 1]
```diff
@@ -350,8 +350,14 @@ void AggregateType::codegenDef() {
...
-        if( largestType ) st = llvm::StructType::get(largestType, NULL);
-        else st = llvm::StructType::get(info->module->getContext());
+        if( largestType ) {
+          st = llvm::StructType::get(largestType
+#if HAVE_LLVM_VER < 50
+                                     , NULL
+#endif
+                                     );
+        } else
+          st = llvm::StructType::get(info->module->getContext());
```
Comment (mppf) [inline compiler/codegen/type.cpp]: This is a trivial/style thing, but I always use curly braces on an else block if I used them on an if block...

File: compiler/include/llvmUtil.h [hunk 2]
```diff
@@ -26,30 +26,7 @@
...
-#elif LLVM_VERSION_MAJOR>3 || (LLVM_VERSION_MAJOR==3 && LLVM_VERSION_MINOR>=4 )
-#define HAVE_LLVM_VER 34
-#elif LLVM_VERSION_MAJOR>3 || (LLVM_VERSION_MAJOR==3 && LLVM_VERSION_MINOR>=3 )
-#define HAVE_LLVM_VER 33
-#elif LLVM_VERSION_MAJOR>3 || (LLVM_VERSION_MAJOR==3 && LLVM_VERSION_MINOR>=2 )
-#define HAVE_LLVM_VER 32
-#elif LLVM_VERSION_MAJOR>3 || (LLVM_VERSION_MAJOR==3 && LLVM_VERSION_MINOR>=1 )
-#define HAVE_LLVM_VER 31
-#endif
+#define HAVE_LLVM_VER (LLVM_VERSION_MAJOR*10 + LLVM_VERSION_MINOR)
```
Comment (mppf) [inline compiler/include/llvmUtil.h]: 👍

File: compiler/util/llvmDebug.cpp [hunk 3]
```diff
@@ -34,10 +34,15 @@
...
-#include "llvm/Support/Dwarf.h"
 #include "llvm/Support/raw_os_ostream.h"
 #include <iostream>
 #include "llvm/ADT/DenseMap.h"
+
+#if HAVE_LLVM_VER >= 50
+#include "llvm/BinaryFormat/Dwarf.h"
+#else
+#include "llvm/Support/Dwarf.h"
+#endif
```
Comment (mppf) [inline compiler/util/llvmDebug.cpp]: Shouldn't this be covered by the include of `llvmDebug.h` above?

File: compiler/codegen/type.cpp (hunk 1 above)
Comment (dmk42) [reply compiler/codegen/type.cpp]: Sounds reasonable.  I will push a fix to make the curly braces the same for the else block.

File: compiler/util/llvmDebug.cpp (hunk 3 above)
Comment (dmk42) [reply compiler/util/llvmDebug.cpp]: Good point.  llvmDebug.cpp includes several files that are already included by llvmDebug.h.  I will push a change to delete the duplicates from llvmDebug.cpp so we don't have to port the files to new versions of LLVM twice.

**Comment by dmk42** [comment]
Style fixes tested with the hellos and `test/llvm`.

**Pull request merged**